doc/align_image_stack.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "ALIGN_IMAGE_STACK ""1"""
.TH ALIGN_IMAGE_STACK "1" "2010-12-26" """Version: 2010.5.0""" "HUGIN"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
align_image_stack \- Align overlapping images for HDR or focus stack creation
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBalign_image_stack\fR [options] \fIinput files\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
align_image_stack is used for the alignment of slightly misaligned images. When several
differently exposed images are taken in the process of creating a High Dynamic Range
image, the images are not perfectly aligned, and the quality of the \s-1HDR\s0 image created
will suffer from this. align_image_stack uses the tools available to hugin to optimize
the roll, pitch and yaw, as well as some parameters of the lens geometry, to align the
images
.SH "OPTIONS"
.IX Header "OPTIONS"
Modes of operation:
.IP "\fB\-p\fR \fIfile\fR" 4
.IX Item "-p file"
Output .pto file (useful for debugging, or further refinement)
.IP "\fB\-a\fR \fIprefix\fR" 4
.IX Item "-a prefix"
Output aligned images as prefix_xxxx.tif
.IP "\fB\-o\fR \fIoutput\fR" 4
.IX Item "-o output"
Align images and merge them to \s-1HDR\s0, generate output.hdr
.PP
Modifiers:
.IP "\fB\-v\fR" 4
.IX Item "-v"
Verbose, print progress messages. Repeat for higher verbosity
.IP "\fB\-e\fR" 4
.IX Item "-e"
Assume input images are full frame fish eye (default: rectilinear)
.IP "\fB\-t\fR \fInum\fR" 4
.IX Item "-t num"
Remove all control points with an error higher than num pixels
(default: 3)
.IP "\fB\-f\fR \fI\s-1HFOV\s0\fR" 4
.IX Item "-f HFOV"
Approximate horizontal field of view of input images, use if \s-1EXIF\s0 info
not complete
.IP "\fB\-m\fR" 4
.IX Item "-m"
Optimize field of view for all images, except for first. Useful for
aligning focus stacks with slightly different magnification.
.IP "\fB\-c\fR \fInum\fR" 4
.IX Item "-c num"
Number of control points (per grid) to create between adjacent images
(default: 8)
.IP "\fB\-l\fR" 4
.IX Item "-l"
Assume linear input files
.IP "\fB\-s\fR \fIscale\fR" 4
.IX Item "-s scale"
Scale down image by 2^scale (default: 1 [2x down\-sampling])
.IP "\fB\-g\fR \fIgsize\fR" 4
.IX Item "-g gsize"
Break image into a rectangular grid (gsize x gsize) and attempt to
find num control points in each section (default: 5 [5x5 grid])
.IP "\fB\-h\fR" 4
.IX Item "-h"
Display help summary.
.SH "AUTHORS"
.IX Header "AUTHORS"
Written by Pablo d'Angelo. Also contains contributions from Douglas Wilkins, Ippei Ukai, Ed Halley, Bruno Postle, Gerry Patterson and Brent Townshend.
.PP
This man page was written by Cyril Brulebois
<cyril.brulebois@enst\-bretagne.fr> and is licensed under the same
terms as the hugin package itself.


NEWS
Newsworthy changes in 0.7 beta 5
--------------------------------
* much improved vignetting and exposure correction,
  http://hugin.sf.net/tech
* ability to create HDR images during stitching
  (still extremly memory hungry, will be improved
   in the next version).
* fixed bad HFOV/VFOV limits of some projections
* OSX build should not crash all the time
* Added "Apply template" function, to apply the
  settings of a previously saved project to
  the current images.
* enabled "load lens data" button in dialog asking
  for focal length/HFOV.

Newsworthy changes in 0.7 beta 4
-------------------------
* update panotools to latest version
* Albers equal area conic and Miller cylindrical projection
* Support for straight line control points
* Fixed running of Autopano-SIFT integration on windows
* (linux) work around bugs in wxWidgets (fixes double deletion
  of control points when pressing the Del key)
* Force even values for width for 360 degree equirectangular
  panoramas.

Newsworthy changes in 0.7 beta 3
-------------------------
* Assistant panel, enables 3 click stitching (well, almost ;-)
* Center and rotate panorama by mouse clicks in preview window.
  Also added a numerical transform dialog
* automatic straighten
* support for Lambert and Lambert Azimuthal projections
* separate crop tab replaces buggy crop dialog
* crop is saved in lens .ini files
* Rotated display of images in control point selector, depending
  on the roll value
* allow images with and without alpha channels in the same project
* option to open final panorama in image editor after stitching
* present output of command line utils in nice window (unix)
* many small bugfixes
* removed image center dialog


Newsworthy changes in 0.6.1
---------------------------
* Autopano-SIFT integration was broken in 0.6 on Unix and OSX.
* Support for displaying HDR files within hugin, using log or gamma mapping
* .hdr file support (tif HDR files have been always useable)
* reduced memory usage, better image caching strategy in GUI.
* fixed crashes and hangs when using the preview auto update and avoid
  unnessecary remapping of images.
* vignetting correction estimation speed up.
* nona and nona_gui can be used to do template stitching, by specifying
  images on the command line (or dnd shortcut under windows)
* distribute fulla.exe with windows package (it was accidentally
  not included in the 0.6 package)
* universal build for Intel Macs, fixed hang after enblend or autopano-sift
  call
* projection can be changed in preview.
* removed all references to malicious panotools.info website
* various smaller bugfixes

Newsworthy changes in 0.6
-------------------------
* updated vigra to version 1.4. All sourcecode of hugin is now
  released under approved open source licences. This means there
  is no problem with including hugin in debian and ubuntu anymore.
* Copy ICC profiles from input images to output
* Added vignetting correction and brighness correction
  during stitching.
  Vignetting can be corrected by addition or division.
  The correction can be done with a flatfield image or an
  radial polynomial
* Functionality to estimate the vignetting polynomial
  directly from the overlapping images. This is an experimental feature.
* Multithreaded remapping. hugin and nona fully utilize multiple
  CPUs or multi core CPUs during preview generation and stitching
* support for dcraw 16 bit .ppm files
* Disabled PTOptimizer progress display on OSX. This should fix
  the crashes some people have experienced.
* New command line tool (fulla), for distortion, TCA and vignetting correction.
  Can use PTLens database (old format only).
* reorganized GUI layout.
* Czech translation added

Newsworthy changes in 0.5
-------------------------
* Added Catalan, Chinese and Hungarian translation
* Support TIFF compressions: NONE, DEFLATE, LZW
* option to create cropped tiff files for usage with enblend 2.4
* Mac OSX: fixed bug during calling PTOptimizer executable
* Windows only:
    hugin is shipped with the pano12.dll build with MSVC.
    This might lead to crashes when using PTStitcher.exe
    In that case replace pano12.dll with pano12_for_usage_with ptstitcher.dll

Newsworthy changes in 0.5 rc2
-----------------------------
* Removed all non-opensource sourcecode from CVS
* Fixed various crashes
* Japanese, Russian, Brazilian and Dutch translation
* Proper default arguments for Autopano-SIFT on Windows
  (users of hugin 0.5 r1 or less need to restore the default
   config)

Newsworthy changes in 0.5 rc1
-----------------------------
* Options on the menu and toolbar that are irrelevant
  are now disabled.
* enblend can now be called from nona even if it has 
  spaces in the path.

Newsworthy changes in 0.5 beta7
-------------------------------
* Updated translations

src/tools/align_image_stack.cpp
// -*- c-basic-offset: 4 -*-

/** @file align_image_stack.cpp
 *
 *  @brief program to align a set of well overlapping images (~90%)
 *
 *  @author Pablo d'Angelo <pablo.dangelo@web.de>
 *
 *  $Id: align_image_stack.cpp 2493 2007-10-24 20:26:26Z dangelo $
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This software is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this software; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include <hugin_config.h>
#include <hugin_version.h>
#include <fstream>
#include <sstream>

#include <vigra/error.hxx>
#include <vigra/impex.hxx>
#include <vigra/cornerdetection.hxx>
#include <vigra/localminmax.hxx>
#include <hugin_utils/utils.h>

#include <vigra_ext/Pyramid.h>
#include <vigra_ext/Correlation.h>
#include <vigra_ext/InterestPoints.h>

#include <panodata/Panorama.h>
#include <panotools/PanoToolsOptimizerWrapper.h>
#include <panodata/StandardImageVariableGroups.h>
#include <algorithms/optimizer/PTOptimizer.h>
#include <nona/Stitcher.h>

#ifdef WIN32
 #include <getopt.h>
#else
 #include <unistd.h>
#endif


#include <tiff.h>

using namespace vigra;
using namespace HuginBase;
using namespace AppBase;
using namespace std;
using namespace vigra_ext;
using namespace HuginBase::PTools;
using namespace HuginBase::Nona;

int g_verbose = 0;

static void usage(const char * name)
{
    cerr << name << ": align overlapping images for HDR creation" << std::endl
         << "align_image_stack version " << DISPLAY_VERSION << std::endl
         << std::endl
         << "Usage: " << name  << " [options] input files" << std::endl
         << "Valid options are:" << std::endl
         << " Modes of operation:" << std::endl
         << "  -p file   Output .pto file (useful for debugging, or further refinement)" << std::endl
         << "  -a prefix align images, output as prefix_xxxx.tif" << std::endl
         << "  -o output merge images to HDR, generate output.hdr" << std::endl
         << " Modifiers" << std::endl
         << "  -v        Verbose, print progress messages.  Repeat for higher verbosity" << std::endl
         << "  -e        Assume input images are full frame fish eye (default: rectilinear)" << std::endl
         << "  -t num    Remove all control points with an error higher than num pixels (default: 3)" << std::endl
         << "  -f HFOV   approximate horizontal field of view of input images, use if EXIF info not complete" << std::endl
         << "  -m        Optimize field of view for all images, except for first." << std::endl
         << "             Useful for aligning focus stacks with slightly different magnification." << std::endl
         << "  -c num    number of control points (per grid) to create between adjacent images (default: 8)" << std::endl
         << "  -l        Assume linear input files" << std::endl
	 << "  -s scale  Scale down image by 2^scale (default: 1 [2x downsampling])" << std::endl
	 << "  -g gsize  Break image into a rectangular grid (gsize x gsize) and attempt to find " << std::endl 
	 << "             num control points in each section (default: 5 [5x5 grid] )" << std::endl
         << "  -h        Display help (this text)" << std::endl
         << std::endl;
}


#if 0
template <class VALUETYPE>
class InterestPointSelector
{
public:

        /** the functor's argument type
        */
    typedef VALUETYPE argument_type;

        /** the functor's result type
        */
    typedef VALUETYPE result_type;

        /** \deprecated use argument_type
        */
    typedef VALUETYPE value_type;


    InterestPointSelector(int nrPoints)
    {
        minResponse = 0;
        nPoints = nrPoints;
    }

    void operator()(argument_type const & resp)
    {
        //double resp = leftCornerResponse(x,y);
	if (resp > minResponse) {
	    // add to point map
	    points.insert(make_pair(resp,Diff2D(x,y)));
	    // if we have reached the maximum
	    // number of points, increase the threshold, to avoid
	    // growing the points map too much.
	    // extract more than nPoints, because some might be bad
	    // and cannot be matched with the other image.
	    if (points.size() > 5*nPoints) {
		// remove the point with the lowest corner response.
		leftCorners(points.begin()->second.x,points.begin()->second.y)=0;
		points.erase(points.begin());
		// use new threshold for next selection.
		minResponse = points.begin()->first;
	    }
	}
    }

    argument_type minResponse;
    std::multimap<argument_type, vigra::Diff2D> points;
    int nPoints;
}

#endif

template <class ImageType>
void createCtrlPoints(Panorama & pano, int img1, const ImageType & leftImg, int img2, const ImageType & rightImg, int pyrLevel, double scale, unsigned nPoints, unsigned grid)
{
    typedef typename ImageType::value_type VT;
    //////////////////////////////////////////////////
    // find interesting corners using harris corner detector
    typedef std::vector<std::multimap<double, vigra::Diff2D> > MapVector;

    std::vector<std::multimap<double, vigra::Diff2D> >points;
    if (g_verbose > 0) {
        std::cout << "Trying to find " << nPoints << " corners... ";
    }

    vigra_ext::findInterestPointsOnGrid(srcImageRange(leftImg, GreenAccessor<VT>()), scale, 5*nPoints, grid, points);

    double scaleFactor = 1<<pyrLevel;

    for (MapVector::iterator mit = points.begin(); mit != points.end(); ++mit) {

        unsigned nGood = 0;
        unsigned nBad = 0;
        // loop over all points, starting with the highest corner score
        for (multimap<double, vigra::Diff2D>::reverse_iterator it = (*mit).rbegin();
             it != (*mit).rend();
             ++it)
        {
            if (nGood >= nPoints) {
                // we have enough points, stop
                break;
            }

            long templWidth = 20;
            long sWidth = 100;
            double corrThresh = 0.9;
            //double curvThresh = 0.0;

            vigra_ext::CorrelationResult res;

            res = vigra_ext::PointFineTune(leftImg,
                                            (*it).second,
                                            templWidth,
                                            rightImg,
                                            (*it).second,
                                            sWidth
                                            );
            if (g_verbose > 2) {
                cout << (*it).second.x << "," << (*it).second.y << " -> "
                        << res.maxpos.x << "," << res.maxpos.y << ":  corr coeff: " <<  res.maxi
                        << " curv:" <<  res.curv.x << " " << res.curv.y << std::endl;
            }
            if (res.maxi < corrThresh )
            {
                nBad++;
                DEBUG_DEBUG("low correlation: " << res.maxi << " curv: " << res.curv);
            } else {
                nGood++;
                // add control point
                ControlPoint p(img1, (*it).second.x * scaleFactor,
                                (*it).second.y * scaleFactor,
                                img2, res.maxpos.x * scaleFactor,
                                res.maxpos.y * scaleFactor);
                pano.addCtrlPoint(p);
            }
        }
        if (g_verbose > 0) {
            cout << "Number of good matches: " << nGood << ", bad matches: " << nBad << std::endl;
        }
    }
};


struct Parameters
{
    Parameters()
    {
        cpErrorThreshold = 3;
        nPoints = 8;
	grid = 5;
        hfov = 0;
        pyrLevel = 1;
	linear = false;	   // Assume linear input files if true
        optHFOV = false;
        fisheye = false;
    }

    double cpErrorThreshold;
    int nPoints;
    int grid;		// Partition images into grid x grid subregions, each with npoints
    double hfov;
    bool linear;
    bool optHFOV;
    bool fisheye;
    int pyrLevel;
    std::string alignedPrefix;
    std::string ptoFile;
    std::string hdrFile;
    string basename;
};

template <class PixelType>
int main2(std::vector<std::string> files, Parameters param)
{
    typedef vigra::BasicImage<PixelType> ImageType;
    try {
        // load first image
        vigra::ImageImportInfo firstImgInfo(files[0].c_str());

        // rescale image
        ImageType * leftImg = new ImageType();
        {
            ImageType leftImgOrig(firstImgInfo.size());
            if(firstImgInfo.numExtraBands() == 1) {
                vigra::BImage alpha(firstImgInfo.size());
                vigra::importImageAlpha(firstImgInfo, destImage(leftImgOrig), destImage(alpha));
            } else if (firstImgInfo.numExtraBands() == 0) {
                vigra::importImage(firstImgInfo, destImage(leftImgOrig));
            } else {
                vigra_fail("Images with multiple extra (alpha) channels not supported");
            }
            reduceNTimes(leftImgOrig, *leftImg, param.pyrLevel);
        }


        Panorama pano;
        Lens l;

        // add the first image.to the panorama object
        // default settings
        double focalLength = 50;
        double cropFactor = 0;
        
        SrcPanoImage srcImg;
        srcImg.setFilename(files[0]);
        
        if (param.fisheye) {
            srcImg.setProjection(SrcPanoImage::FULL_FRAME_FISHEYE);
        }
        srcImg.readEXIF(focalLength, cropFactor, true, true);
        // disable autorotate
        srcImg.setRoll(0);
        if (srcImg.getSize().x == 0 || srcImg.getSize().y == 0) {
            cerr << "Could not decode image: " << files[0] << "Unsupported image file format";
            return 1;
        }

        // use hfov specified by user.
        if (param.hfov > 0) {
            srcImg.setHFOV(param.hfov);
        } else if (cropFactor == 0) {
            // could not read HFOV, assuming default: 50
            srcImg.setHFOV(50);
        }
        
        if (param.linear) {
            srcImg.setResponseType(SrcPanoImage::RESPONSE_LINEAR);
            if (g_verbose>0) {
                cout << "Using linear response" << std::endl;
            }
        }
        
        pano.addImage(srcImg);
        
        // setup output to be exactly similar to input image
        PanoramaOptions opts;

        if (param.fisheye) {
            opts.setProjection(PanoramaOptions::FULL_FRAME_FISHEYE);
        } else {
            opts.setProjection(PanoramaOptions::RECTILINEAR);
        }
        opts.setHFOV(srcImg.getHFOV(), false);

        if (srcImg.getRoll() == 0.0 || srcImg.getRoll() == 180.0) {
            opts.setWidth(srcImg.getSize().x, false);
            opts.setHeight(srcImg.getSize().y);
        } else {
            opts.setWidth(srcImg.getSize().y, false);
            opts.setHeight(srcImg.getSize().x);
        }
            // output to tiff format
        opts.outputFormat = PanoramaOptions::TIFF_m;
        opts.tiff_saveROI = false;
        // m estimator, to be more robust against points on moving objects
        opts.huberSigma = 2;
        pano.setOptions(opts);

        // variables that should be optimized
        // optimize nothing in the first image
        OptimizeVector optvars(1);

        ImageType * rightImg = new ImageType(leftImg->size());
        StandardImageVariableGroups variable_groups(pano);

        // loop to add images and control points between them.
        for (int i = 1; i < (int) files.size(); i++) {
            if (g_verbose > 0) {
                cout << "Creating control points between " << files[i-1] << " and " << files[i] << std::endl;
            }
            // add next image.
            srcImg.setFilename(files[i]);
            srcImg.readEXIF(focalLength, cropFactor, true, true);
            if (srcImg.getSize().x == 0 || srcImg.getSize().y == 0) {
                cerr << "Could not decode image: " << files[i] << "Unsupported image file format";
                return 1;
            }
            if (param.hfov > 0) {
                srcImg.setHFOV(param.hfov);
            } else if (cropFactor == 0) {
                // could not read HFOV, assuming default: 50
                srcImg.setHFOV(50);
            }
            
            int imgNr = pano.addImage(srcImg);
            variable_groups.update();
            // each image shares the same lens.
            variable_groups.getLenses().switchParts(imgNr, 0);
            // unlink HFOV?
            if (param.optHFOV) {
                pano.unlinkImageVariableHFOV(0);
            }
            
            // All images are in the same stack: Link the stack variable.
            pano.linkImageVariableStack(imgNr, 0);
            
            // load the actual image data of the next image
            vigra::ImageImportInfo nextImgInfo(files[i].c_str());
            assert(nextImgInfo.size() == firstImgInfo.size());
            {
                ImageType rightImgOrig(nextImgInfo.size());
                if (nextImgInfo.numExtraBands() == 1) {
                    vigra::BImage alpha(nextImgInfo.size());
                    vigra::importImageAlpha(nextImgInfo, destImage(rightImgOrig), destImage(alpha));
                } else if (nextImgInfo.numExtraBands() == 0) {
                    vigra::importImage(nextImgInfo, destImage(rightImgOrig));
                } else {
                    vigra_fail("Images with multiple extra (alpha) channels not supported");
                }
                reduceNTimes(rightImgOrig, *rightImg, param.pyrLevel);
            }

            // add control points.
            // work on smaller images
            // TODO: or use a fast interest point operator.
            createCtrlPoints(pano, i-1, *leftImg, i, *rightImg, param.pyrLevel, 2, param.nPoints, param.grid);

            // swap images;
            delete leftImg;
            leftImg = rightImg;
            rightImg = new ImageType(leftImg->size());

            // optimize yaw, roll, pitch
            std::set<std::string> vars;
            vars.insert("y");
            vars.insert("p");
            vars.insert("r");
            if (param.optHFOV) {
                vars.insert("v");
            }
            optvars.push_back(vars);
        }
        delete leftImg;
        delete rightImg;

        // optimize everything.
        pano.setOptimizeVector(optvars);
        bool optimizeError = false;
        optimizeError = (PTools::optimize(pano) > 0);

        // need to do some basic outlier pruning.
        // remove all points with error higher than a specified threshold
        if (param.cpErrorThreshold > 0) {
            CPVector cps = pano.getCtrlPoints();
            CPVector newCPs;
            for (int i=0; i < (int)cps.size(); i++) {
                if (cps[i].error < param.cpErrorThreshold) {
                    newCPs.push_back(cps[i]);
                }
            }
            if (g_verbose > 0) {
                cout << "Ctrl points before pruning: " << cps.size() << ", after: " << newCPs.size() << std::endl;
            }
            pano.setCtrlPoints(newCPs);
            // reoptimize
            optimizeError = (PTools::optimize(pano) > 0) ;
        }

        UIntSet imgs = pano.getActiveImages();

        if (param.ptoFile.size() > 0) {
            std::ofstream script(param.ptoFile.c_str());
            pano.printPanoramaScript(script, optvars, pano.getOptions(), imgs, false, "");
        }

        if (optimizeError)
        {
            cerr << "An error occured during optimization." << std::endl;
            cerr << "Try adding \"-p debug.pto\" and checking output." << std::endl;
            cerr << "Exiting..." << std::endl;
            return 1;
        }

        if (param.hdrFile.size()) {
            // TODO: photometric alignment (HDR, fixed white balance)
            //utils::StreamProgressReporter progress(2.0);
            //loadImgsAndExtractPoints(pano, nPoints, pyrLevel, randomPoints, progress, points);
            //smartOptimizePhotometric

            // switch to HDR output mode
            PanoramaOptions opts = pano.getOptions();
            opts.outputFormat = PanoramaOptions::HDR;
            opts.outputPixelType = "FLOAT";
            opts.outputMode = PanoramaOptions::OUTPUT_HDR;
            opts.outfile = param.hdrFile;
            pano.setOptions(opts);

            // remap all images
            StreamMultiProgressDisplay progress(cout);
            stitchPanorama(pano, pano.getOptions(),
                           progress, opts.outfile, imgs);
        }
        if (param.alignedPrefix.size()) {
            // disable all exposure compensation stuff.
            PanoramaOptions opts = pano.getOptions();
            opts.outputExposureValue = 0;
            opts.outputMode = PanoramaOptions::OUTPUT_LDR;
            opts.outputFormat = PanoramaOptions::TIFF_m;
            opts.outputPixelType = "";
            opts.outfile = param.alignedPrefix;
            pano.setOptions(opts);
            for (unsigned i=0; i < pano.getNrOfImages(); i++) {
                SrcPanoImage img = pano.getSrcImage(i);
                img.setExposureValue(0);
                pano.setSrcImage(i, img);
            }
            // remap all images
            StreamMultiProgressDisplay progress(cout);
            stitchPanorama(pano, pano.getOptions(),
                           progress, opts.outfile, imgs);

        }
    } catch (std::exception & e) {
        cerr << "ERROR: caught exception: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}

int main(int argc, char *argv[])
{
    // parse arguments
    const char * optstring = "a:ef:g:hlmp:vo:s:t:c:";
    int c;

    opterr = 0;

    g_verbose = 0;

    Parameters param;
//    // use to override exposure value on the command line?
//    std::map<std::string, double> exposureValueMap;
    while ((c = getopt (argc, argv, optstring)) != -1)
        switch (c) {
        case 'a':
            param.alignedPrefix = optarg;
            break;
        case 'c':
            param.nPoints = atoi(optarg);
	    if (param.nPoints<1) {
		cerr << "Invalid parameter: Number of points/grid (-c) must be at least 1" << std::endl;
		return 1;
	    }
            break;
        case 'e':
            param.fisheye = true;
            break;
        case 'f':
            param.hfov = atof(optarg);
	    if (param.hfov<=0) {
		cerr << "Invalid parameter: HFOV (-f) must be greater than 0" << std::endl;
		return 1;
	    }
            break;
	case 'g':
	    param.grid = atoi(optarg);
	    if (param.grid <1 || param.grid>50) {
		cerr << "Invalid parameter: number of grid cells (-g) should be between 1 and 50" << std::endl;
		return 1;
	    }
	    break;
	case 'l':
	    param.linear = true;
	    break;
        case 'm':
            param.optHFOV = true;
            break;
        case 't':
            param.cpErrorThreshold = atof(optarg);
	    if (param.cpErrorThreshold <= 0) {
		cerr << "Invalid parameter: control point error threshold (-t) must be greater than 0" << std::endl;
		return 1;
	    }
            break;
        case 'p':
            param.ptoFile = optarg;
            break;
        case 'o':
            param.hdrFile = optarg;
            break;
        case 'v':
            g_verbose++;
            break;
        case 'h':
            usage(argv[0]);
            return 0;
	case 's':
	    param.pyrLevel = atoi(optarg);
	    if (param.pyrLevel<0 || param.pyrLevel >8) {
		cerr << "Invalid parameter: scaling (-s) should be between 0 and 8" << std::endl;
		return 1;
	    }
	    break;
        default:
            cerr << "Invalid parameter: " << optarg << std::endl;
            usage(argv[0]);
            return 1;
        }

    unsigned nFiles = argc - optind;
    if (nFiles < 2) {
        std::cerr << std::endl << "Error: at least two files need to be specified" << std::endl <<std::endl;
        usage(argv[0]);
        return 1;
    }

    if (param.hdrFile.size() == 0 && param.ptoFile.size() == 0 && param.alignedPrefix.size() == 0) {
        std::cerr << std::endl
                  << "ERROR: Please specify at least one of the -p, -o or -a options." << std::endl
                  << std::endl;
        usage(argv[0]);
        return 1;
    }

    // extract file names
    std::vector<std::string> files;
    for (size_t i=0; i < nFiles; i++)
        files.push_back(argv[optind+i]);

    // TODO: sort images in pano by exposure

    std::string pixelType;

    try {
        vigra::ImageImportInfo firstImgInfo(files[0].c_str());
        pixelType = firstImgInfo.getPixelType();
    } catch (std::exception & e) {
        cerr << "ERROR: caught exception: " << e.what() << std::endl;
        return 1;
    }

    if (pixelType == "UINT8") {
        return main2<RGBValue<UInt8> >(files, param);
    } else if (pixelType == "INT16") {
        return main2<RGBValue<Int16> >(files, param);
    } else if (pixelType == "UINT16") {
        return main2<RGBValue<UInt16> >(files, param);
    } else if (pixelType == "FLOAT") {
        return main2<RGBValue<float> >(files, param);
    } else {
        cerr << " ERROR: unsupported pixel type: " << pixelType << std::endl;
        return 1;
    }

    return 0;
}
